### DISCUSSION TITLE
Minimizing startup time

### DISCUSSION THREAD

**Discussion by rocketraman** [graphql-java/graphql-java#2452] [Q&A]
I'm trying to minimize application startup time (serverless environment).

I'm finding about 75% of my startup time is currently dealing with GraphQL initialization. Approximately half of that is generating the `GraphQLSchema` via [graphql-kotlin](https://github.com/ExpediaGroup/graphql-kotlin) and the other half of that is `GraphQL.newGraphQL(graphQLSchema).build()`.

What are my options to reduce my startup time? Currently GraphQL initialization takes about 3 seconds -- I'd like to get it down to less than half a second if I can. One thought I had was a way to serialize a `GraphQL` object once at build-time, and deserialize it at runtime, but the object is not Serializable.

**Comment by bbakerman** [comment]
This is a great question - one that we don't have a great answer on.  Most of our efforts have been on making queries cheap and not so much on the schema build.  There is a lot of rules in a valid type system but we have not spent lots of time making it faster

@andimarek had been discussing this with me of late.

If you are an a AWS lambda world, you can cheat by having objects that actually live longer than the Lambda invocation

And since a schema tends to be shared across all users, then it's safe to squirrel away.

But this is a work around to the fact that Graphql Schema objects  an be expensive to build.

The larger the number of elements in the schema, the greater the time it takes to create.  if you willing to share your schema with us, we would love to access to it.

We have a https://github.com/graphql-java/graphql-anonymizer tool that can be used to take a schema (printed) and create an anonymised version.

**Reply by rocketraman** [reply]
Thanks for your reply @bbakerman.

> The larger the number of elements in the schema, the greater the time it takes to create.

Understood. A common approach when something takes a long time to create, and that time cannot be optimized away, is to build it at compile time and squirrel it away for runtime. Is this an avenue you guys have been considering?

> if you willing to share your schema with us, we would love to access to it.

Here is the anonymized schema: https://gist.github.com/rocketraman/065a4a523a094e4e2c0438f4f304c5db. This is going to grow significantly as well -- we are still relatively early in the development of this project.

> If you are an a AWS lambda world, you can cheat by having objects that actually live longer than the Lambda invocation

I'm using Cloud Run, not AWS. As far as I know the only solution there is to specify minimum instances as 1, so that there is always one warm started instance. This has cost implications, and it doesn't solve the cold start issue when scaling to handle spiky loads.

**Reply by bbakerman** [reply]
See 

PR https://github.com/graphql-java/graphql-java/pull/2453

PR  https://github.com/graphql-java/graphql-java/pull/2454

**Comment by bbakerman** [comment]
> and the other half of that is GraphQL.newGraphQL(graphQLSchema).build().

This I find VERY surprising.   This is a simple object allocation and build - there is no code other that field sets here that I can see.

I would love to see more perf information on this.

**Reply by rocketraman** [reply]
> This I find VERY surprising. This is a simple object allocation and build - there is no code other that field sets here that I can see.

You're right -- my timer was setup incorrectly, and I was basically double-counting the time taken to generate `GraphQLSchema` via graphql-kotlin. The graphql-kotlin code [1] to create `GraphQLSchema` is taking almost 2 seconds -- the call to `GraphQL.newGraphQL(schema).build()` is taking only 12 ms.

However, that 1.5s is still huge, and only going to get longer. Overall, if I want to experiment with reducing my schema generation time, I was initially thinking a mechanism in or with graphql-kotlin to cache the results of the schema generation -- but the authors there suggested [2] they can't really help with that and that's something graphql-java has to support. I see PR #2454 makes `TypeDefinitionRegistry` serializable but as far as I can tell graphql-kotlin doesn't use TypeDefinitionRegistry (see [1]) so I'm not sure if that helps me.

[1] https://github.com/ExpediaGroup/graphql-kotlin/blob/master/generator/graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/SchemaGenerator.kt

[2] https://github.com/ExpediaGroup/graphql-kotlin/issues/1214

**Reply by rocketraman** [reply]
Yes, I've confirmed making `TypeDefinitionRegistry` does not help my use case because graphql-kotlin generates the types from code, and does not use `TypeDefinitionRegistry`.

What I'd really need is a way to serialize and restore `GraphQLSchema` itself.

**Comment by bbakerman** [comment]
So I think graphql-kotlin used a code first approach - that is you define a
schema in code not SDL strings

So it seems that their "discovery of the appropriate code" is really slow,
especially if  GraphQL.newGraphQL(schema).build() is taking only 12 ms.

> What I'd really need is a way to serialize and restore GraphQLSchema
 itself.

nominally you can print it to string and parse it back in - we found
parsing is 50/50 on schema gen so that is like 24 ms.  HOWEVER your in
trouble because the backing DataFetcher code is now probably all wrong and
wired up incorrectly.

You might want to move away from code -> schema if its too slow

**Reply by bbakerman** [reply]
Looking into graphql=kotlin - it uses class scanning to find classes that it should turn into graphql types etc..

There are fast class scanners out there (and the one it uses https://github.com/classgraph/classgraph is fast) however with enough class scanning its ALWAYS going to be slower than NOT doing that especially if you have a big schema

Yours is a 500 line schema which is while not tiny is not small.  So I would suggest the Kotlin class scanning approach here is problematic.

Which in a classic compute environment is not a problem - the cost is amortised over many requests - but in a  serverless world - its bad

**Comment by knotenpunkt** [comment]
@bbakerman some more performance question here: To get the data from pojos in datafetcher results, you use dynamic java reflection. Does it make sense to speed this up too? Maybe to create on compile time special classes, which access the pojo, so that we dont need the  relfecation on runtime?

**Comment by bbakerman** [comment]
The Property DataFetcher does use reflection. Reflection in Java is not as
slow as some people think.  Slower than direct method access but not an
order of magnitude slower.

Also it has a JVM wide cache to the methods so that once it has been
called, its holds the methods needed for that data fetcher and hence its as
fast a code method invocation.

Your challenge is that in a Serverless world, you are not amortising any of
the costs across requests.

You can however write your own data fetchers that called `pojo.getName()`
direct.  This is more than just allowing a `name` field ot be a
reflected PropertyDataFetcher call but its faster.

The question I have is - does graphql-lkotlin allow you to inject your own
data fetchers - I dont know
